wolfcrypt/src/eccsi.c

```diff
@@ -0,0 +1,2209 @@
+ * @return  BAD_FUNC_ARG when key or data is NULL.
+ * @return  BUFFER_E when size of data is not equal to the expected size.
+ * @return  MP_MEM or MEMORY_E when dynamic memory allocation fails.
+ */
+int wc_ImportEccsiPrivateKey(EccsiKey* key, const byte* data, word32 sz)
+{
+    int err = 0;
+
+    if ((key == NULL) || (data == NULL)) {
+        err = BAD_FUNC_ARG;
+    }
+    if ((err == 0) && (sz != (word32)key->ecc.dp->size)) {
+        err = BUFFER_E;
+    }
+
+    if (err == 0) {
+        err = mp_read_unsigned_bin(&key->ecc.k, data, key->ecc.dp->size);
+    }
+
+    return err;
+}
+
+/**
+ * Export the KMS Public Auth Key (KPAK) from the ECCSI object.
+ *
+ * KPAK is required by all clients in order to perform cryptographic operations.
+ *
+ * X and y ordinate of public key concatenated. Each number is zero padded to
+ * key size.
+ * Descriptor byte (0x04) is prepeneded when not raw.
+ *
+ * @param  [in]      key      ECCSI key.
+ * @param  [out]     data     Buffer to hold the encoded public key.
+ * @param  [in,out]  sz       On in, size of buffer in bytes.
+ *                            On out, length of encoded public key in bytes.
+ * @param  [in]      raw   On 0, prepend descriptor byte.
+ *                         On 1, only include ordinates.
+ * @return  0 on success.
+ * @return  BAD_FUNC_ARG when key or sz is NULL.
+ * @return  LENGTH_ONLY_E when data is NULL - sz is set.
+ * @return  BUFFER_E when the buffer passed in is too small.
+ */
+int wc_ExportEccsiPublicKey(EccsiKey* key, byte* data, word32* sz, int raw)
+{
+    int err = 0;
+
+    if ((key == NULL) || (sz == NULL)) {
+        err = BAD_FUNC_ARG;
+    }
+    if ((err == 0) && (key->ecc.type != ECC_PRIVATEKEY) &&
+            (key->ecc.type != ECC_PUBLICKEY)) {
+        err = BAD_STATE_E;
+    }
+
+    if ((err == 0) && (data != NULL)) {
+        err = eccsi_kpak_from_mont(key);
+    }
+    if (err == 0) {
+        /* Write out public key. */
+        err = eccsi_encode_point(&key->ecc.pubkey, (word32)key->ecc.dp->size,
+            data, sz, raw);
+    }
+
+    return err;
+}
+
+/*
+ * Generates an (SSK, PVT) Pair - signing key pair.
+ *
+ * RFC 6507, Section 5.1.1
+ *
+ * @param  [in]   key       ECCSI key.
+ * @param  [in]   rng       Random number generator.
+ * @param  [in]   hashType  Type of hash algorithm. e.g. WC_SHA256
+ * @param  [in]   id        Identity to create hash from.
+ * @param  [in]   idSz      Length of identity in bytes.
+ * @param  [out]  ssk       Secret Signing Key as an MP integer.
+ * @param  [out]  pvt       Public Validation Token (PVT) as an ECC point.
+ * @return  0 on success.
+ * @return  MP_MEM or MEMORY_E when dynamic memory allocation fails.
+ * @return  Other -ve value when an internal operation fails.
+ */
+static int eccsi_make_pair(EccsiKey* key, WC_RNG* rng,
+        enum wc_HashType hashType, const byte* id, word32 idSz, mp_int* ssk,
+        ecc_point* pvt)
+{
+    int err;
+    byte hashSz = 0;
+
+    do {
```

Comment:
Should we have a counter and exit this loop if fails lots of times in a row to prevent an infinite loop?

Reply:
Added.
Not in spec but good to have!